### PR TITLE
catalog,optimize: Cluster-specific `Optimize` pipelines

### PR DESCRIPTION
**Pull request by aalexandrov**
Part 2 of #25273.

### Motivation

  * This PR adds a known-desirable feature.

### Tips for reviewer

- The first two commits wire the necessary changes in the catalog.
- The third commit wires the overrides in the `OptimizerConfig`. I'll see if this part can be done in the `optimizer_feature_flags` macro in a follow-up PR.
- The fourth commit adds an `*.slt` test for the new functionality.
- The final commit adds a new `OptimizerConfig` option (see commit message for details).

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/pull/25010md).
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - There are no user-facing behavior changes.

### CODE REVIEW

**Review by mjibson — APPROVED** [review]
Adapter parts lgtm.

**Comment by aalexandrov** [comment]
@jkosh44 there is [a failing nightly test in the platform checks](https://buildkite.com/materialize/nightlies/builds/6604#018de1e6-51b9-44b9-8632-abcade4096b4) that fails with this message:

```
platform-checks-materialized-1      | thread 'main' panicked at src/catalog/src/durable/impls/shadow.rs:508:48: assertion `left == right` failed
```

but my migration script are identical for both cases. Any idea what might be causing this?

File: src/adapter/src/optimize/mod.rs [hunk 1]
```diff
@@ -151,7 +152,34 @@ where
...
 
-// Feature flags for the optimizer.
+/// Feature flags for the optimizer.
+///
+/// To add a new feature flag, do the following steps:
+///
+/// 1. To make the flag available to all stages in our [`Optimize`] pipelines:
+///    1. Add the flag as an [`OptimizerConfig`] field.
+///
+/// 2. To allow users to set a system-wide override for this feature flag:
```
Comment (ggevay) [inline src/adapter/src/optimize/mod.rs:162]: "To allow users"
It's not users who do this, but we do it through LD, right?

File: src/adapter/src/optimize/mod.rs [hunk 2]
```diff
@@ -151,7 +152,34 @@ where
...
+///    3. Extend the `TryFrom<ExplainPlanOptionExtracted>` implementation for
+///       [`mz_repr::explain::ExplainConfig`].
+///    4. Extend the `OverrideFrom<ExplainContext>` implementation for
+///       [`OptimizerConfig`].
+///
+/// 4. To enable `CLUSTER ... FEATURES(...)` overrides:
+///    1. Add the flag to the `optimizer_feature_flags!(...)` macro call.
+///    2. Add the flag to the `impl_simple_options!(ClusterFeature {...})` macro
+///       call.
+///    2. Extend the `let optimizer_feature_overrides = ...` call in
```
Comment (ggevay) [inline src/adapter/src/optimize/mod.rs:179]: 3.

File: src/adapter/src/coord/sequencer/inner/create_index.rs [hunk 3]
```diff
@@ -228,6 +228,7 @@ impl Coordinator {
             self.allocate_transient_id()?
         };
         let optimizer_config = optimize::OptimizerConfig::from(self.catalog().system_config())
+            .override_from(&self.catalog.get_cluster(*cluster_id).config.features())
```
Comment (ggevay) [inline src/adapter/src/coord/sequencer/inner/create_index.rs:231]: Should we wire this up also in `subscribe.rs`? (maybe also copy_to?)

File: test/sqllogictest/cluster_features.slt
Comment (ggevay) [inline test/sqllogictest/cluster_features.slt:1]: Some ideas for more tests:
- Different statement types
  - index and peek
  - EXPLAIN CREATE ...
  - EXPLAIN REPLAN
- Overriding the cluster's default in an EXPLAIN statement.
- ALTER CLUSTER

File: test/sqllogictest/cluster_features.slt [hunk 4]
```diff
@@ -19,3 +19,138 @@ mode cockroach
...
+  t1.y as c1,
+  t2.y as c2,
+  t3.y as c3
+FROM
+  t1, t2, t3
+where
+  t1.x = t2.x AND
+  t2.y = t3.y;
+
+# EXPLAIN mv2 on c2 (should be running with the feature flag turned on).
```
Comment (ggevay) [inline test/sqllogictest/cluster_features.slt:96]: mv1 on c1

File: src/adapter/src/optimize/mod.rs [hunk 5]
```diff
@@ -151,7 +152,34 @@ where
...
+///       `plan_create_cluster`.
+///    4. Extend the `OverrideFrom<OptimizerFeatureOverrides>` implementation
+///       for [`OptimizerConfig`].
```
Comment (ggevay) [inline src/adapter/src/optimize/mod.rs:182]: You could also mention potentially adding
```
self.reoptimize_imported_views = true;
```
to `impl OverrideFrom<OptimizerFeatureOverrides> for OptimizerConfig` and to `impl OverrideFrom<ExplainContext> for OptimizerConfig`.

File: src/adapter/src/optimize/dataflows.rs [hunk 6]
```diff
@@ -308,7 +308,7 @@ impl<'a> DataflowBuilder<'a> {
         df_desc: &mut DataflowDesc,
         config: &OptimizerConfig,
     ) -> Result<(), OptimizerError> {
-        if config.mode == OptimizeMode::Explain {
+        if config.reoptimize_imported_views {
```
Comment (ggevay) [inline src/adapter/src/optimize/dataflows.rs:311]: Update doc comment of `reoptimize_imported_views`.

Comment (ggevay) [reply src/adapter/src/optimize/dataflows.rs:311]: Also, maybe rename to something like `maybe_reoptimize_imported_views`.

**Review by ggevay — COMMENTED** [review]
I wrote some minor comments.

(There is a Nightly assert panic, but I have no idea what that assert is checking.)

File: src/adapter/src/coord/sequencer/inner/create_index.rs (hunk 3 above)
Comment (aalexandrov) [reply src/adapter/src/coord/sequencer/inner/create_index.rs:232]: Added to `subscribe.rs`. The `COPY TO` pipeline piggy-backs on `inner/peek.rs` so it's already handled.

File: src/adapter/src/optimize/mod.rs (hunk 5 above)
Comment (aalexandrov) [reply src/adapter/src/optimize/mod.rs:182]: Discussed offline. I've changed the commit to promote `reoptimize_imported_views` to a feature flag.

**Comment by jkosh44** [comment]
> @jkosh44 there is [a failing nightly test in the platform checks](https://buildkite.com/materialize/nightlies/builds/6604#018de1e6-51b9-44b9-8632-abcade4096b4) that fails with this message:
> 
> ```
> platform-checks-materialized-1      | thread 'main' panicked at src/catalog/src/durable/impls/shadow.rs:508:48: assertion `left == right` failed
> ```
> 
> but my migration script are identical for both cases. Any idea what might be causing this?

For the record, this was fixed in https://github.com/MaterializeInc/materialize/pull/25593

**Review by ggevay — APPROVED** [review]

**Pull request merged**